api/logs/s3/s3.go

```diff
@@ -85,26 +124,51 @@ func path(appName, callID string) string {
 }
 
 func (s *store) InsertLog(ctx context.Context, appName, callID string, callLog io.Reader) error {
-	objectName := path(appName, callID)
-	_, err := s.client.PutObjectWithContext(ctx, s.bucket, objectName, callLog, -1, minio.PutObjectOptions{ContentType: contentType})
-	return err
-}
+	span, ctx := opentracing.StartSpanFromContext(ctx, "s3_insert_log")
+	defer span.Finish()
+
+	// wrap original reader in a decorator to keep track of read bytes without buffering
+	cr := &countingReader{r: callLog}
 
-func (s *store) GetLog(ctx context.Context, appName, callID string) (io.Reader, error) {
 	objectName := path(appName, callID)
-	obj, err := s.client.GetObjectWithContext(ctx, s.bucket, objectName, minio.GetObjectOptions{})
+	params := &s3manager.UploadInput{
+		Bucket:      aws.String(s.bucket),
+		Key:         aws.String(objectName),
+		Body:        cr,
+		ContentType: aws.String(contentType),
+	}
+
+	logrus.WithFields(logrus.Fields{"bucketName": s.bucket, "key": objectName}).Debug("Uploading log")
+	_, err := s.uploader.UploadWithContext(ctx, params)
 	if err != nil {
-		return nil, err // this is always nil, for now, thanks minio :(
+		return fmt.Errorf("failed to write log, %v", err)
 	}
 
-	_, err = obj.Stat()
+	span.LogFields(log.Int("fn_s3_log_upload_size", cr.count))
+	return nil
+}
+
+func (s *store) GetLog(ctx context.Context, appName, callID string) (io.Reader, error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "s3_get_log")
+	defer span.Finish()
+
+	objectName := path(appName, callID)
+	logrus.WithFields(logrus.Fields{"bucketName": s.bucket, "key": objectName}).Debug("Downloading log")
+
+	// stream the logs to an in-memory buffer
+	target := &aws.WriteAtBuffer{}
```

Comment:
have used https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.GetObject w/ some success in the past to avoid having to copy into a buffer here, am less familiar with the `s3Manager` api.

Reply:
happy to punt on optimizing here

Reply:
I think that should work as well. Will look at making the change tomorrow.

Reply:
@rdallman I've looked into replacing this with `s3.GetObject` and the challenge there is that we wouldn't be able to simply return the Body of `s3.GetObjectOutput` as the io.Reader result of `GetLog`. The reason is that the result body needs to be closed in order to not leak connections (see https://github.com/aws/aws-sdk-go/issues/408), so unfortunately we wouldn't be able to avoid copying into an intermediate buffer.

Reply:
gotcha. we could change `LogStore.GetLog` to return an `io.ReadCloser` ?